Make sanitizer special case list slash-agnostic

DKLoehr opened the pull request:
This changes the glob matcher for the sanitizer special case format so that it treats `/` as matching both forward and back slashes.

When dealing with cross-compiles or build systems that don't normalize slashes, it's possible to run into file paths with inconsistent slashiness, e.g. `../..\v8/include\v8-internal.h` when [building chromium](https://g-issues.chromium.org/issues/425364464).

We can match this using the current syntax using this ugly kludge: `src:*{/,\\}v8{/,\\}*`. However, since the format is explicitly for listing file paths, it makes sense to treat `/` as denoting a path separator rather than a literal forward slash. This allows us to write the much more natural form `src:*/v8/*` and have it work on any platform.

This is technically a behavior change, but it seems very unlikely to come up in practice. It will only make a difference if a user has a system where they want to catch `a/b` but not `a\b`. Even in the worst case, they can still regain the previous behavior by escaping the slash character in their pattern: `src:a\/b`.

Note: this is not default behavior yet. To activate use `#!special-case-list-v4` in the first line of the special case list file.

DKLoehr committed:
Make special case matcher slash-agnostic

DKLoehr committed:
Enable only for special case list

DKLoehr commented:
> GlobPattern is also used in binary utilities' symbol/section name matching where we want to be rigid

Thanks for pointing that out! I narrowed the scope of the change so that the parts of the code which want it have to explicitly enable it. Right now that's just the special case list, which is the target of these changes in any case.

DKLoehr commented:
Friendly ping on this

MaskRay commented:
You might consider adding more reviewers to gather broader perspectives on this feature. I'm neutral about it personally.

jyknight commented on llvm/lib/Support/SpecialCaseList.cpp:
```diff
@@ -59,7 +59,8 @@ Error SpecialCaseList::Matcher::insert(StringRef Pattern, unsigned LineNumber,
   // reference which could be destroyed before match() is called
-  if (auto Err = GlobPattern::create(Glob->Name, /*MaxSubPatterns=*/1024)
+  if (auto Err = GlobPattern::create(Glob->Name, /*MaxSubPatterns=*/1024,
+                                     /*IsSlashAgnostic=*/true)
```
In particular: this should be true only when built for Windows.

DKLoehr replied:
Is it possible there are cases during a cross-compilation when we might see varied slashiness for different hosts/targets?

jyknight replied:
A filename with an actual backslash in it is valid on Unix, and would be incorrectly matched. (Of course, the likelihood of someone using a source file with such an odd filename is low.)

DKLoehr replied:
Ah, that's very sad to hear. Thanks for explaining it to me.

DKLoehr committed:
Enable only on windows

DKLoehr committed:
Incorporate review feedback

DKLoehr commented:
@jyknight Friendly ping on this.

vitalybuka reviewed (commented):
I'd like to optimize SanitizerSpecialCaseList so we sort by prefixes/suffixes to limit applied matches.

IsSlashAgnostic will be needed to taken into account soring.
Which is possible.

Would be simple just normalize patters on load?
Same for queries?

vitalybuka reviewed (changes requested):
I believe this patch will not handle pattern:
path "clang\something" agains patter "src:clang/".

Can you please extend the test?

vitalybuka commented:
There is GlobPatternTest,cpp which should be extended.

DKLoehr committed:
Add glob pattern unittest

vitalybuka commented on llvm/lib/Support/GlobPattern.cpp:
```diff
@@ -137,7 +137,7 @@ GlobPattern::create(StringRef S, std::optional<size_t> MaxSubPatterns) {
 
   // Store the prefix that does not contain any metacharacter.
-  size_t PrefixSize = S.find_first_of("?*[{\\");
+  size_t PrefixSize = S.find_first_of("?*[{\\/");
```
Now it breaks optimization for all platforms.

Maybe add a member of GlobPattern::IsSlashAgnostic?

DKLoehr replied:
> Maybe add a member of GlobPattern::IsSlashAgnostic?

This was the original strategy, IIRC I changed it because we wanted to restrict it only to filenames and that information wasn't readily available when we're making it a glob. I suppose we could thread it through if we have to.

vitalybuka commented:
> Would be simple just normalize patters on load? Same for queries?

WDYT?

DKLoehr commented:
> Would be simple just normalize patters on load? Same for queries?

I don't know what the implications of this are, but it seems much simpler so we can try that instead.

I don't know what a query is in this context.

DKLoehr committed:
Canonicalize special case list filenames when loading

vitalybuka commented:
> > Would be simple just normalize patters on load? Same for queries?
> 
> I don't know what the implications of this are, but it seems much simpler at this point so we can try that instead.
> 
> I don't know what a query is in this context.

It's query parameter `SpecialCaseList::Matcher::match(StringRef Query)`

But it needs to be normalized on higher levels, when we know this is path.

Annoying part is that to normalize globs you need to '\' -> '/'.
But to do that you needs some glob parsing to distinguish escape sequence \ and real \ .

And then I don't know which one will be better.



NOTE: There is very simple option, no sure how acceptable.
1. Special list always exact / (for windows and posix)
2. 'query' on windows replace \ to / in file name.

This one must be easy to do.

vitalybuka commented:
> I've been experimenting with the simple option (canonicalize file names before matching, and require the patterns to use `/` as file separators on all systems), but it requires anyone who uses a case list with backslashes to rewrite it, which I suspect would be considered too onerous to merge.
> 

Seems like /\ ambiguity is also error prone. It's glob, how often \ will be confused with escape sequence?

> If we don't impose that requirement, then we'd need to edit the globs as well, but I'm hesitant to mess with user-provided patterns because it seems easy to produce unintuitive behavior (and makes debugging harder on their end), e.g. they wonder why their `\` in a pattern isn't matching anything.

You have a difference problem now, why "\\" does not match as expected (what ever is user expected)?

Given that, I'd prefer we just make windows also rely on /.

DKLoehr commented:
> how to match "foo\[bar]" ?

I'm not sure if I'm interpreting the question correctly, but  if you're asking about a file named `[bar]` in a directory `foo` on a windows, then:
1. With LLVM trunk, you would write `src:foo\\\[bar]` (I think you don't need to escape the `]`)
2. With the current PR, you would write `src:foo/\[bar]`, OR `src:foo\\\[bar]` if you know the path will be windows-style (it will fail to match unix-style paths)
3. If we canonicalized file names before matching them, you would have to write `src:foo/\[bar]` (`src:foo\\\[bar]` would not work, and you'd have to update your special case files to use the new syntax)
4. If we "canonicalized" globs, you could write either `src:foo/\[bar]` OR `src:foo\\\[bar]`, but the latter would also (incorrectly) match unix-style paths. This is ok as long as you're canonicalizing your paths, but if you apply it to something that isn't a path (or which shouldn't be canonicalized for some reason) then it's wrong.

Basically, if you manipulate the globs you're effectively adding a new rule that `\\` matches either `/` or `\`, but that contradicts with the existing rule that it should only match exactly `\`.

I'm entirely ok with option (3); it's simple to implement, simple to state how it works, and doesn't make the glob parser slower or more complicated. The downside is it puts burden on downstream folks to change their files to the new syntax.

That burden might not be that large, because it only applies to files which are windows-specific. If you support multiple platforms, you'll have to use a workaround anyway (in chromium we use `{\\,/}` to match either separator). So only a subset of people will actually be impacted, but I don't know how to evaluate that so I'd like to get opinions from more experienced community members.

vitalybuka commented on clang/lib/Basic/Diagnostic.cpp:
```diff
@@ -595,10 +612,18 @@ bool WarningsSpecialCaseList::isDiagSuppressed(diag::kind DiagId,
-        CategoriesToMatchers,
-        llvm::sys::path::remove_leading_dotslash(PLoc.getFilename()));
+  if (PresumedLoc PLoc = SM.getPresumedLoc(DiagLoc); PLoc.isValid()) {
+    if (CanonicalizePaths) {
```
is going to be common thing for all users of 
```
if (CanonicalizePaths) {
      return globsMatches(
          CategoriesToMatchers,
          llvm::sys::path::convert_to_slash(
              llvm::sys::path::remove_leading_dotslash(PLoc.getFilename())));
    } else {
      return globsMatches(
          CategoriesToMatchers,
          llvm::sys::path::remove_leading_dotslash(PLoc.getFilename()));
    }
```

Lets put that thing into `SpecialCaseList::NormalizePath()`

vitalybuka commented on clang/docs/SanitizerSpecialCaseList.rst:
```diff
@@ -196,6 +196,23 @@ tool-specific docs.
+  stripping any leading dots and slashes, and (on Windows only) converting all
+  backslashes to forward slashes.
+
+  If the file uses both ``#!special-case-list-v1`` and ``#!canonical-paths``,
```
And we should list flags which uses special case list

jyknight commented:
I'm sorry I haven't yet read all the recent comments regarding technical issues with supporting both / and \, but just want to say my initial impression is that it'd be a massive overkill to force an incompatible format-change here, instead of just supporting both separators on Windows. I'll write a more informed reply later.

vitalybuka commented:
> Keeping backward compatibility in ignore lists looks like unnecessary maintenance burden. Migration is easy, but if we not doing this, we stuck with that for years.

BTW. We want to do another breaking change in this version.

```
src:./path/
src:path/
```

Warning file does this righght and removes ./,
I'd like to expand this on all lists.

So we can bundle those changes them together.

vitalybuka commented:
I just realized we missed a trivial solution:

**I think it's reasonable to expect that a single pattern does not mix \ / as a path separator in the same pattern.**

So, on windows we match the path as is, and then also match the converted version.

Maybe: As a bonus we can add a warning suggesting to convert a suppression file into forward slashes. So in distant future we can keep only / versions.

nico commented:
> I think it's reasonable to expect that a single pattern does not mix \ / as a path separator in the same pattern.

Eh, if you imagine that a tool generates the file, then it's pretty easy to end up with mixes slashiness, since `os.path.join` might use `\` to join parts and manually-written fragments might use `/`. If we're going the extra mile here, we might as well go it all the way?

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
Resolve merge conflicts

vitalybuka committed:
[SpecialCaseList] Replace #!canonical-paths with #!special-case-list-v4

vitalybuka committed:
Revert "[SpecialCaseList] Canonicalize paths in Matcher::match"

vitalybuka committed:
[Support] Remove CanonicalizePaths member from SpecialCaseList

vitalybuka committed:
[Support] Rename CanonicalizePaths to CanonicalizeSlashes

vitalybuka committed:
Update docs

vitalybuka committed:
windows only

vitalybuka committed:
format

vitalybuka committed:
[Support] Move CanonicalizeSlashes after RemoveDotSlash

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
[Support] Disallow backslashes in path patterns for SpecialCaseList v4

vitalybuka committed:
Revert "[Support] Disallow backslashes in path patterns for SpecialCaseList v4"

This reverts commit dddc25166eb422cf9b54cbc52bff2c1d628db53a.

vitalybuka committed:
format

vitalybuka committed:
Docs

vitalybuka committed:
[NFC][Docs] Revert changes to llvm/docs/ReleaseNotes.md

These changes were originally added for the canonicalized paths design, which was abandoned in favor of slash-agnostic matching. The correct documentation is now in clang/docs/ReleaseNotes.rst.

Assisted-by: Gemini

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
[NFC][Support] Refactor bracket parsing in GlobPattern

Extract BitVector from Expected before flipping to
prepare for future modifications.

Pull Request: https://github.com/llvm/llvm-project/pull/202848

vitalybuka committed:
[NFC][Support] Refactor bracket parsing in GlobPattern

Extract BitVector from Expected before flipping to
prepare for future modifications.

Pull Request: https://github.com/llvm/llvm-project/pull/202848

vitalybuka committed:
[NFC][Support] Define Prefix/SuffixMetacharacters constants

Extract literal metacharacter strings used in GlobPattern into static constexpr arrays to improve consistency and maintainability.

Assisted-by: Gemini

Pull Request: https://github.com/llvm/llvm-project/pull/202850

vitalybuka committed:
[NFC][Support] Define Prefix/SuffixMetacharacters constants

Extract literal metacharacter strings used in GlobPattern into static constexpr arrays to improve consistency and maintainability.

Assisted-by: Gemini

Pull Request: https://github.com/llvm/llvm-project/pull/202850

vitalybuka committed:
[Support] Implement slash-agnostic path matching in GlobPattern

Add a SlashAgnostic option to GlobPattern to allow matching path separators
(both forward slashes and backslashes) agnostically.

When enabled:
- We conservatively reduce the plain prefix and suffix by treating path
  separators as metacharacters. This ensures that path separators are
  matched via the slash-agnostic state machine rather than plain string
  comparison.
- Brackets containing slashes are adjusted to match both separators.
- Character comparisons in the state machine (matchChar) treat '/' and
  '\' as equivalent.

Co-authored-by: Devon Loehr <DKLoehr@users.noreply.github.com>

Assisted-by: Gemini

Pull Request: https://github.com/llvm/llvm-project/pull/202855

vitalybuka committed:
[Support] Implement slash-agnostic path matching in GlobPattern

Add a SlashAgnostic option to GlobPattern to allow matching path separators
(both forward slashes and backslashes) agnostically.

When enabled:
- We conservatively reduce the plain prefix and suffix by treating path
  separators as metacharacters. This ensures that path separators are
  matched via the slash-agnostic state machine rather than plain string
  comparison.
- Brackets containing slashes are adjusted to match both separators.
- Character comparisons in the state machine (matchChar) treat '/' and
  '\' as equivalent.

Co-authored-by: Devon Loehr <DKLoehr@users.noreply.github.com>

Assisted-by: Gemini

vitalybuka committed:
[Support] Implement slash-agnostic path matching in GlobPattern

Add a SlashAgnostic option to GlobPattern to allow matching path separators
(both forward slashes and backslashes) agnostically.

When enabled:
- We conservatively reduce the plain prefix and suffix by treating path
  separators as metacharacters. This ensures that path separators are
  matched via the slash-agnostic state machine rather than plain string
  comparison.
- Brackets containing slashes are adjusted to match both separators.
- Character comparisons in the state machine (matchChar) treat '/' and
  '\' as equivalent.

Co-authored-by: Devon Loehr <DKLoehr@users.noreply.github.com>

Assisted-by: Gemini

Pull Request: https://github.com/llvm/llvm-project/pull/202854

vitalybuka committed:
Merge branch 'glob' into slashy

vitalybuka committed:
fixed

vitalybuka committed:
Merge branch 'glob' into slashy

vitalybuka committed:
Merge branch 'glob' into slashy

vitalybuka committed:
[NFC][Support] Use is_style_windows in SpecialCaseList

Use llvm::sys::path::is_style_windows(llvm::sys::path::Style::native) instead of is_separator('\\') to check if native path style is Windows.

Assisted-by: Gemini

vitalybuka committed:
[NFC][Support] Use is_style_windows in SpecialCaseList

Use llvm::sys::path::is_style_windows(llvm::sys::path::Style::native) instead of is_separator('\\') to check if native path style is Windows.

Assisted-by: Gemini

vitalybuka committed:
Merge commit 'd19dcda89e1a217e07efb0f0899e76c480b82c40' into slashy

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
resolve

vitalybuka committed:
MinVersion

vitalybuka committed:
fixup! MinVersion

vitalybuka committed:
fixup! [SpecialCaseList] Replace #!canonical-paths with #!special-case-list-v4

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
format

MaskRay commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -1029,6 +1029,12 @@ Sanitizers
   requires rules to match canonicalized paths (without leading ``./``).
 
```
no need for blank lines

MaskRay commented on llvm/lib/Support/SpecialCaseList.cpp:
```diff
@@ -406,6 +415,15 @@ bool SpecialCaseList::parse(unsigned FileIdx, const MemoryBuffer *MB,
   bool UseGlobs = MinVersion(2);
   bool RemoveDotSlash = MinVersion(3);
   bool WarnDotSlash = MinVersion(4) && !MinVersion(5);
+  // FIXME: Improve efficiency on Windows.
```
Inefficiency should use TODO instead of FIXME

vitalybuka committed:
Merge remote-tracking branch 'origin/main' into slashy

vitalybuka committed:
Address review comments: Remove blank lines and use TODO instead of FIXME
